lang/localize-core-element.js

```diff
@@ -0,0 +1,71 @@
+import { LocalizeMixin } from '../mixins/localize-mixin.js';
+
+export const LocalizeCoreElement = superclass => class extends LocalizeMixin(superclass) {
+
+	static async getLocalizeResources(langs) {
+		let translations;
+		for await (const lang of langs) {
+			switch (lang) {
+				case 'ar':
+					translations = await import('./ar.js');
+					break;
+				case 'da':
+					translations = await import('./da.js');
+					break;
+				case 'de':
+					translations = await import('./de.js');
+					break;
+				case 'en':
+					translations = await import('./en.js');
+					break;
+				case 'es':
+					translations = await import('./es.js');
+					break;
+				case 'fr':
+					translations = await import('./fr.js');
+					break;
+				case 'ja':
+					translations = await import('./ja.js');
+					break;
+				case 'ko':
+					translations = await import('./ko.js');
+					break;
+				case 'nl':
+					translations = await import('./nl.js');
+					break;
+				case 'pt':
+					translations = await import('./pt.js');
+					break;
+				case 'sv':
+					translations = await import('./sv.js');
+					break;
+				case 'tr-tr':
+					translations = await import('./tr-tr.js');
+					break;
+				case 'tr':
+					translations = await import('./tr.js');
+					break;
+				case 'zh-cn':
+					translations = await import('./zh-cn.js');
+					break;
+				case 'zh-tw':
+					translations = await import('./zh-tw.js');
+					break;
+				case 'zh':
+					translations = await import('./zh.js');
+					break;
+			}
+			if (translations && translations.default) {
```

Comment:
The static mixin would automatically assign values for missing translations by iterating over the languages in reverse order.  With this new way, how are missing translations handled, say at dev-time, and prod?

Reply:
ah I see how that works - yeah with this current implementation the expectation would be that the translations are complete for each language and so it would just get the first language that has a translations file. I like that idea. It would increase the number of requests (I think up to 4, though that would generally be unlikely and usually 2-3)

Reply:
Besides the extra requests, another downside to what was being done was - it was doing that object merge per instance of the component.  I think we had a store to cache them.

It would be nice to avoid that runtime processing, and also the extra requests?  What are people's thoughts around doing some type of validation to ensure we don't publish with missing translations?

Reply:
Personally if the extra requests are only helpful during dev... I'd rather not do them. Seems silly for all our users to make 2-3 extra requests forever for no reason just to make dev a bit easier.

I'd be pro some kind of validation around catching missing translations -- there's already some Serge-based validation I believe that tries to catch replace strings that are messed up.

Reply:
For validation, what are thoughts there? My initial thought was adding a script that runs after build when travis is deploying (run it here https://github.com/BrightspaceUI/core/blob/master/.travis.yml#L54 - it could also run as part of the tests stage) which checks if all the lang files have the same keys. It could/probably should also check if the values are empty, though that would be problematic right now since we are missing some translations (should we add english values which would then get translated?).

Reply:
I think the deployment stage runs on all builds, ya?  It just doesn't actually deploy to NPM unless it's for a tag.

So... do we want this validation on every build?  and do we want to block PRs if those language files are missing translations?

I think if we go that route, we would manually populate missing values with English while we are waiting for Serge - hopefully that would not cause problems with Serge.  I imagine Serge would eventually just overwrite those values.

Another option is to hook into NPM's `prepublishOnly` step to validate and/or populate missing translation values just for the code being published.  This has the advantage of working in the event that the package needs to be published outside of CI.  It leaves the translation files in the source unmodified, which is good in the sense that we aren't doctoring the source files, but bad in the sense that there would be missing translation values while dev'ing on core - which is perhaps not a big deal.

Reply:
So just to confirm I understand what using `prepublishOnly` would do, is that we would run a script that would replace missing translations with some value which would then go into the publish code, but would not be in the source code? If this is the case, would we still want to have some sort of alert to whoever published?

Reply:
Yeah that's what the idea would be if we want to hook into `prepublishOnly`.  

We could write a message into the log but I suspect it would almost always go unnoticed.  I can't speak for everyone, but I usually only look if CI failed, I suspect NPM publish didn't work, or I am checking which goldens were updated.

I tend to lean that way, but I don't feel super strongly about it.  It also makes it a bit easier to identify untranslated terms when looking at the code.

If we were to go the other route, and populate English values for missing translations and check them into GitHub while we're waiting for the Serge process, would those values cause problems for Serge, or will it still identify them as being untranslated and post an update for them?

Reply:
Based on the my courses PR request here https://github.com/Brightspace/d2l-my-courses-ui/pull/821/files it looks like if we use English values they will be replaced with translations. @svanherk was there anything extra needed to have translations done that way or was that just automatically how it worked?
If that's the case then I think it would probably just be easiest to populate with English values pre-translation.

Reply:
The way I believe it works is:
1. Serge will look to see if you've added or changed any English terms.
2. It will send only those terms to the translators, regardless of whether you have those terms translated already or not.
3. Serge will come back with the translated terms and open a PR.

Because we're setting serge up for the first time, all the English terms we have will be sent.  We might get back some different terms for the ones we already translated, and the English terms in other languages should be replaced with the correct translated ones.

So sort answer: Yes, I think you're good to populate untranslated terms with the English versions and serge will properly replace them

Reply:
My only reservation with populating in our source code is that it makes it harder for us to identify untranslated terms thereafter.  My suspicion is that the logic for validating is almost identical to populating, and the timing of when it gets called doesn't really introduce any difference - it just gets tacked onto existing script vs adding a line to package.json for `prepublishOnly`.

Reply:
okay yeah that makes sense to me. I'll work on adding a script to run as part of `prepublishOnly`